ansible (now) works with python3, but the ansible backend does not

see: https://docs.ansible.com/ansible/python_3_support.html

but when I try to run testinfra --connection=ansible, I get output that looks like this:

```
$ testinfra --connection=ansible -v
=== test session starts ===
platform darwin -- Python 3.6.0, pytest-3.0.7, py-1.4.33, pluggy-0.4.0 -- /Users/me/.virtualenvs/infra/bin/python3.6
cachedir: .cache
rootdir: /Users/me/src/group/infra, inifile:
plugins: testinfra-1.6.0
collected 0 items / 1 errors

=== ERRORS ===
___ ERROR collecting test_CI.py ___
../../../.virtualenvs/infra/lib/python3.6/site-packages/testinfra/plugin.py:136: in pytest_generate_tests
    ansible_inventory=metafunc.config.option.ansible_inventory,
../../../.virtualenvs/infra/lib/python3.6/site-packages/testinfra/host.py:125: in get_hosts
    for backend in testinfra.backend.get_backends(hosts, **kwargs):
../../../.virtualenvs/infra/lib/python3.6/site-packages/testinfra/backend/__init__.py:84: in get_backends
    for name in klass.get_hosts(host, **kw):
../../../.virtualenvs/infra/lib/python3.6/site-packages/testinfra/backend/ansible.py:82: in get_hosts
    return AnsibleRunner(kwargs.get("ansible_inventory")).get_hosts(host)
../../../.virtualenvs/infra/lib/python3.6/site-packages/testinfra/utils/ansible_runner.py:139: in __init__
    _reload_constants()
../../../.virtualenvs/infra/lib/python3.6/site-packages/testinfra/utils/ansible_runner.py:47: in _reload_constants
    reload(ansible.constants)
E   NameError: name 'reload' is not defined
=== pytest-warning summary ===
WP1 None Module already imported so can not be re-written: testinfra
!!! Interrupted: 1 errors during collection !!!
=== 1 pytest-warnings, 1 error in 0.75 seconds ===
```

here's a proposed diff/fix:

```
--- a/testinfra/backend/ansible.py
+++ b/testinfra/backend/ansible.py
@@ -47,12 +47,12 @@ class AnsibleBackend(base.BaseBackend):
         # A simple test case is:
         # >>> assert File("/bin/true").content == open("/bin/true").read()
         try:
-            stdout_bytes = b"".join((chr(ord(c)) for c in out['stdout']))
+            stdout_bytes = out['stdout'].encode('utf-8')
         except ValueError:
             stdout_bytes = None

         try:
-            stderr_bytes = b"".join((chr(ord(c)) for c in out['stderr']))
+            stderr_bytes = out['stderr'].encode('utf-8')
         except ValueError:
             stderr_bytes = None
--- a/testinfra/utils/ansible_runner.py
+++ b/testinfra/utils/ansible_runner.py
@@ -16,7 +16,6 @@
 from __future__ import unicode_literals
 from __future__ import absolute_import

+from imp import reload
 import pprint

 try:
```


I've only been able to get the tests partially working, but it seems like this should happen too:
```
--- a/testinfra/test/test_backends.py
+++ b/testinfra/test/test_backends.py
@@ -12,7 +12,6 @@
 # limitations under the License.
 from __future__ import unicode_literals
 import pytest
-import six

 import testinfra.backend
 BACKENDS = ("ssh", "safe-ssh", "docker", "paramiko", "ansible")
@@ -86,8 +85,6 @@ def test_backend_importables():
     # just check that all declared backend are importable and NAME is set
     # correctly
     for connection_type in testinfra.backend.BACKENDS:
-        if six.PY3 and connection_type == 'ansible':
-            pytest.skip()
         obj = testinfra.backend.get_backend_class(connection_type)
         assert obj.get_connection_type() == connection_type
```

Haven't tested the `.encode` bit on python2 either

--- a/test/test_backends.py
+++ b/test/test_backends.py
@@ -12,7 +12,6 @@
 # limitations under the License.
 from __future__ import unicode_literals
 import pytest
-import six
 
 import testinfra.backend
 BACKENDS = ("ssh", "safe-ssh", "docker", "paramiko", "ansible")
@@ -86,8 +85,6 @@ def test_backend_importables():
     # just check that all declared backend are importable and NAME is set
     # correctly
     for connection_type in testinfra.backend.BACKENDS:
-        if six.PY3 and connection_type == 'ansible':
-            pytest.skip()
         obj = testinfra.backend.get_backend_class(connection_type)
         assert obj.get_connection_type() == connection_type
 

--- a/testinfra/backend/ansible.py
+++ b/testinfra/backend/ansible.py
@@ -19,6 +19,7 @@ import pprint
 
 from testinfra.backend import base
 from testinfra.utils.ansible_runner import AnsibleRunner
+from testinfra.utils.ansible_runner import to_bytes
 
 logger = logging.getLogger("testinfra")
 
@@ -42,27 +43,16 @@ class AnsibleBackend(base.BaseBackend):
     def run(self, command, *args, **kwargs):
         command = self.get_command(command, *args)
         out = self.run_ansible("shell", module_args=command)
-
-        # Ansible may return bytes as an unicode object...
-        # A simple test case is:
-        # >>> assert File("/bin/true").content == open("/bin/true").read()
-        try:
-            stdout_bytes = b"".join((chr(ord(c)) for c in out['stdout']))
-        except ValueError:
-            stdout_bytes = None
-
-        try:
-            stderr_bytes = b"".join((chr(ord(c)) for c in out['stderr']))
-        except ValueError:
-            stderr_bytes = None
-
         return self.result(
             out['rc'],
             command,
-            stdout_bytes=stdout_bytes,
-            stderr_bytes=stderr_bytes,
+            stdout_bytes=None,
+            stderr_bytes=None,
             stdout=out["stdout"], stderr=out["stderr"],
         )
+
+    def encode(self, data):
+        return to_bytes(data)
 
     def run_ansible(self, module_name, module_args=None, **kwargs):
         result = self.ansible_runner.run(

--- a/testinfra/utils/ansible_runner.py
+++ b/testinfra/utils/ansible_runner.py
@@ -16,7 +16,9 @@
 from __future__ import unicode_literals
 from __future__ import absolute_import
 
+import imp
 import pprint
+
 
 try:
     import ansible
@@ -40,11 +42,19 @@ elif _ansible_major_version == 2:
     import ansible.utils.vars
     import ansible.vars
 
+try:
+    from ansible.module_utils._text import to_bytes
+except ImportError:
+    from ansible.utils.unicode import to_bytes
+
+
+__all__ = ['AnsibleRunner', 'to_bytes']
+
 
 def _reload_constants():
     # Reload defaults that can depend on environment variables and
     # current working directory
-    reload(ansible.constants)
+    imp.reload(ansible.constants)
 
 
 class AnsibleRunnerBase(object):

--- a/tox.ini
+++ b/tox.ini
@@ -4,7 +4,7 @@ envlist=py27,ansible1,py34,flake8,pylint,sphinxdoc
 [testenv]
 deps=
     -rtest-requirements.txt
-    py27: ansible>=2.3,<3
+    ansible>=2.3,<3
 commands=
     testinfra {posargs:-v -n 4 --cov testinfra --cov-report xml --cov-report term test}
 usedevelop=True

